### ISSUE TITLE
Gorouter: Negroni used in Gorouter (last updated 2018) blocks us from using newer Golang features, e.g. `ResponseController`.

### ISSUE THREAD

**Issue by peanball** [cloudfoundry/routing-release#373]
No

## Issue

Go 1.20 and 1.21 introduced improved support for full duplex processing of HTTP requests for HTTP/1.1. Our use of the `Negroni` middleware handler is blocking experiments with this.

~~Negroni has not been updated since 2018. We might need to fork or replace it in order to keep receiving improvements or rather extended functionality from the Go standard library.~~

Update, thanks to @mariash: there are v2 and v3 versions of negroni with recent updates. 

The `ResponseController` specifically may enable better communication error handling with backends, earlier detection that the connection is in fact not established or earlier error responses. ~~We cannot easily check without effort of extending Negroni.~~

v3 of negroni adds an `Unwrap()` method that provides access to the raw response, enabling support for the `ResponseController`.
## Affected Versions

All.

## Traffic Diagram 

N/A

## Steps to Reproduce
N/A
## Expected result

We can use the `ResponseController` and enable `EnableFullDuplex`.

## Current result

The Negroni handler does not expose `EnableFullDuplex` and is not suitable for use with the ResponseController without modifications.

## Possible Fix

Modify our vendored Negroni and expose `EnableFullDuplex`, and run the experiments. If this is successful we may need to fork Negroni.

Alternatively, we can consider replacing Negroni with a different and maintained middleware framework.

## Additional Context

@geofffranks, please add more detail as you see fit.

**Comment by geofffranks** [comment]
I wonder if we could swap out negroni for something like https://github.com/go-chi/chi. It makes use of standard net/http interfaces instead of its own versions, and seems to be actively maintained.

**Comment by geofffranks** [comment]
If we were to do this, we'd want to try to do it as an experimental opt-in implementation with the new multiplexer at first to iron out any bugs and make sure we understand the performance differences between the two.

**Comment by maxmoehl** [comment]
If all we need is middleware we could just do that ourselves:
```go
type Middleware interface {
	Register(http.Handler) http.Handler
}

// Chain all middleware in the order provided. The final middleware will call
// the provided handler.
func Chain(h http.Handler, ms ...Middleware) http.Handler {
	// Reverse order since we have to register the last handler first.
	for i := len(ms) - 1; i >= 0; i++ {
		h = ms[i].Register(h)
	}

	return h
}

// Func is a convenience method to turn a function into a middleware.
type Func func(http.Handler) http.Handler

func (m Func) Register(h http.Handler) http.Handler {
	return m(h)
}
```

Using it would look like this:

```go
var Slog = Func(func(h http.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		slog.Info("received http request", "method", r.Method) // ...
		h.ServeHTTP(w, r)
		slog.Info("request done")
	})
})

type contextKeyVal struct {
	key, val string
}

func NewContextKeyVal(key, val string) Middleware {
	return contextKeyVal{key: key, val: val}
}

func (m contextKeyVal) Register(h http.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		r.WithContext(context.WithValue(r.Context(), m.key, m.val))
		h.ServeHTTP(w, r)
	})
}

func main() {
	mw := []Middleware{
		Slog,
		NewContextKeyVal("foo", "bar"),
	}

	// our final handler which we obtained somehow
	var h http.Handler

	h = Chain(h, mw...)
}
```

**Comment by peanball** [comment]
I think this is a bit of a simplistic view. While it is nice to have control and be closer to the Go standard library, this means re-implementing many things that other middleware frameworks already provide out of the box.

Looking at the handlers already available for frameworks such as chi, there is much to be reused, and some handlers that we often thought would be nice to also have in Gorouter.

Before creating our own framework, we should give existing ones an honest look first.

**Comment by maxmoehl** [comment]
I agree, if there is a library / framework that provides more useful features that we'd like to use, we should use that. But _just_ for middleware I wouldn't pull in a library as we do usually require very tight control over the request flow and building such a helper is fairly simple.

**Comment by mariash** [comment]
Commented here: https://github.com/cloudfoundry/gorouter/pull/380#issuecomment-1949392320

> @peanball so I looked into this more and negroni provides an Unwrap method to the ResponseWriter - https://github.com/urfave/negroni/commit/59d32aab419da1606c459c848c7586ef52b07a4b which will satisfy the latest ResponseController and eventually EnableFullDuplex will be called on the underlying http.ResponseWriter.
> 
> Negroni actually has v2 and v3 versions, that are not visible on their releases page, but can be imported as github.com/urfave/negroni/v3. The problem is that Unwrap was not released on v3 yet. And looking at https://github.com/urfave/negroni/issues/274 the last comment from December 2023 stated that author is temporarily unavailable. So I would give it couple more months to see if he will be back and cut a new release. For now I pinned the negroni to latest v3.
> 
> Here is the proposed change for the Gorouter: https://github.com/cloudfoundry/gorouter/pull/395
> Related change in routing-release: https://github.com/cloudfoundry/routing-release/pull/389

**Comment by peanball** [comment]
I've adapted the description to reflect those newest findings

**Comment by mariash** [comment]
Hi @peanball negroni released 3.1.0 with `Unwrap` method on `ResponseController` and we started using it in [this commit of the Gorouter](https://github.com/cloudfoundry/gorouter/commit/f0b4d8effe240fc66fa7ac20b19a582abce41d9d). Going to close this issue. Thank you!
